keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,364 @@
+a primary communication path from the control plane to the nodes is from the apiserver to the kubelet process running on each node.
+The kubelet communicates with the container runtime over gRPC, where kubelet is the gRPC client and the Container Runtime Interface is the gRPC server.
+The CRI then sends the creation request to the container runtime installed on the node.
+Traces gathered from the kubelet will provide critical information to monitor and troubleshoot interactions at the node level.
+This KEP proposes using OpenTelemetry libraries, and exports in the OpenTelemetry format.
+This is in line with the [API Server enhancement](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing).
+
+## Motivation
+
+Along with metrics and logs, traces are a useful form of telemetry to aid with debugging incoming requests.
+The kubelet can make use of distributed tracing to improve the ease of use and enable easier analysis of trace data.
+Trace data is structured, providing the detail necessary to debug requests across service boundaries.
+As more core components are instrumented, Kubernetes becomes easier to monitor, manage, and troubleshoot.
+
+### Definitions
+
+**Span**: The smallest unit of a trace.  It has a start and end time. Spans are the building blocks of a trace.    
+**Trace**: A collection of Spans which represents work being done by a system. A record of the path of requests through a system.    
+**Trace Context**: A reference to a Trace that is designed to be propagated across component boundaries.    
+
+### Goals
+
+* The kubelet generates and exports spans for reconcile loops it initiates and for outgoing requests to the CRI. An example of a reconcile loop is the
+creation of a pod. Pod creation involves pulling the image, creating the pod sandbox and creating the container. Those would be linked together into a
+single trace to diagnose latency issues within that flow.
+
+### Non-Goals
+
+* Tracing in other kubernetes controllers besides the kubelet
+* Replace existing logging, metrics
+* Change metrics or logging (e.g. to support trace-metric correlation)
+* Access control to tracing backends
+* Add tracing to components outside kubernetes (e.g. etcd client library).
+
+## Proposal
+
+### User Stories
+
+Since this feature is for diagnosing problems with the kubelet, it is targeted at cluster operators, administrators, and cloud vendors that 
+manage kubernetes nodes and core services.
+
+For the following use-cases, I can deploy an OpenTelemetry collector agent as a DaemonSet to collect kubelet trace data from each node's host network.
+From there, OpenTelemetry trace data can be exported to a tracing backend of choice.  I can use the kubelet's `--opentelemetry-config-file flag` with
```

Comment:
prefer `--tracing-config-file`, as opentelemetry is scoped to more than tracing, but we only currently plan to use it for tracing.

Reply:
I prefer --tracing-config-file, also. However, I think the best way forward is to add the TracingConfiguration to kubernetes/compent-base/config, as you suggest. If it's decided to use the flag instead, it will be `tracing-config-file`